### PR TITLE
Add dparray

### PR DESCRIPTION
**Pull request by PokhodenkoSA**
This code is from #184 placed in common branch.

From https://github.com/IntelPython/dpctl/pull/192#issuecomment-739069862:

- [ ] Doc-strings for the new class and the new module should be added.

### CODE REVIEW

**Comment by vlad-perevezentsev** [comment]
Fixed comments @oleksandr-pavlyk :
https://github.com/IntelPython/dpctl/pull/184#pullrequestreview-541258709
https://github.com/IntelPython/dpctl/pull/184#pullrequestreview-541259450
https://github.com/IntelPython/dpctl/pull/184#pullrequestreview-541260020

Warning and error @PokhodenkoSA (https://github.com/IntelPython/dpctl/pull/184#issuecomment-735957700 ) did not find after making changes. 

Commands for the conda virtual environment:
conda create -n dpctl-dev python=3.7 -c http://fxsatmac02.an.intel.com:8091
conda install cython -c http://fxsatmac02.an.intel.com:8091
conda install numpy -c http://fxsatmac02.an.intel.com:8091
conda install cmake -c http://fxsatmac02.an.intel.com:8091
export ONEAPI_ROOT=/opt/intel/oneapi
source /opt/intel/oneapi/compiler/latest/env/vars.sh
python setup.py develop

File: setup.py [hunk 1]
```diff
@@ -201,6 +201,7 @@ def _get_cmdclass():
     cmdclass["develop"] = develop
     return cmdclass
 
+print("packages:", find_packages(include=["*"]))
```
Comment (PokhodenkoSA) [inline setup.py:204]: Forgot to delete?

File: dpctl/__init__.pxd [hunk 2]
```diff
@@ -29,4 +29,4 @@
 
 from dpctl._sycl_core cimport *
 from dpctl._memory import *
-
+from dparray import *
```
Comment (PokhodenkoSA) [inline dpctl/__init__.pxd:32]: As I know it is an absolute import. How it is possible to import dparray which is inside dpctl package via absolute path?
```suggestion
from .dparray import *
```

File: dpctl/tests/test_dparray.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+def multiply_func(np_array, dpcrtl_array):
+    return np_array * dpcrtl_array
+
+
+class TestOverloadList(unittest.TestCase):
+    maxDiff = None
+
+    X = dparray.ndarray((256, 4), dtype='d')
+    X.fill(1.0)
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_dparray.py:18]: This variables will be shared among test cases. It means that cases are not independent.

File: dpctl/dparray.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+debug = False
+
+def dprint(*args):
+    if debug:
+        print(*args)
+        sys.stdout.flush()
+
+import dpctl
+from dpctl.memory import MemoryUSMShared
```
Comment (PokhodenkoSA) [inline dpctl/dparray.py:18]: Imports usually are on top of the file. Is it a specific case?

**Comment by PokhodenkoSA** [comment]
Please, format code with black formatter.

File: dpctl/dparray.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
+import numpy as np
+from inspect import getmembers, isfunction, isclass, isbuiltin
+from numbers import Number
+from types import FunctionType as ftype, BuiltinFunctionType as bftype
+import sys
+#import importlib
+#import functools
```
Comment (PokhodenkoSA) [inline dpctl/dparray.py:7]: It is better to remove commented code.

File: dpctl/tests/test_dparray.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        self.assertIsInstance(C, dparray.ndarray)
+
+    def test_dparray_through_python_func(self):
+        C = self.X * 5
+        dp_func = func_operation_with_const(C)
+        self.assertIsInstance(dp_func, dparray.ndarray)
+
+    def test_dparray_mixing_dpctl_and_numpy(self):
+        dp_numpy = numpy.ones((256, 4), dtype='d')
+        res = multiply_func(dp_numpy, self.X)
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_dparray.py:47]: This function is used once. I think test will be more clear if this function will be replaced with its body.

File: dpctl/tests/test_dparray.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        dp1 = dparray.from_ndarray(Y)
+        self.assertIsInstance(dp1, dparray.ndarray)
+
+    def test_multiplication_dparray(self):
+        C = self.X * 5
+        self.assertIsInstance(C, dparray.ndarray)
+
+    def test_dparray_through_python_func(self):
+        C = self.X * 5
+        dp_func = func_operation_with_const(C)
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_dparray.py:42]: This function could be expanded. I do not see any reason to use function here.

File: dpctl/__init__.pxd [hunk 8]
```diff
@@ -29,4 +29,4 @@
 
 from dpctl._sycl_core cimport *
 from dpctl._memory import *
-
+from .dparray import *
```
Comment (diptorupd) [inline dpctl/__init__.pxd:32]: Why is this needed in the `__init__.pxd`? It does not really belong here.

Comment (DrTodd13) [reply dpctl/__init__.pxd:32]: Maybe this is how he got dparray to be visible if the user does dpctl.dparray.

File: dpctl/dparray.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        # When called from the explicit constructor, obj is None
+        if obj is None:
+            return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:122]: To wean `dpctl` off dependency on `numba` we need to rely on `obj` to provide `__sycl_usm_array_interface__` and make this the only way we detect that `obj` has the underlying USM memory.

We will also need to validate the content of the `__sycl_usm_array_interface__` making sure that the USM pointer is meaningful in the context that we work with.

Doing this will require changes in Numba-DPPY, and possibly Numba itself, which are queued up.

File: dpctl/dparray.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                dprint(dir(mobj))
+                if ea == dppl_rt_allocator:
+                    return
+                mobj = mobj.parent
+                if isinstance(mobj, ndarray):
+                    mobj = mobj.base
+        if isinstance(obj, np.ndarray):
+            ob = self
+            while isinstance(ob, np.ndarray):
+                if hasattr(obj, array_interface_property):
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:145]: This needs to be `hasattr(ob, array_interface_property)`.

File: dpctl/dparray.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+    # Convert to a NumPy ndarray.
+    def as_ndarray(self):
+        return np.copy(self)
+
+    def __array__(self):
+        return self
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        if method == "__call__":
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:167]: Need to also handle `method="accumulate"` , `"reduce"`, `"reduceat"`, `"outer"`

File: dpctl/dparray.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+def dprint(*args):
+    if debug:
+        print(*args)
+        sys.stdout.flush()
+
+
+functions_list = [o[0] for o in getmembers(np) if isfunction(o[1]) or isbuiltin(o[1])]
+class_list = [o for o in getmembers(np) if isclass(o[1])]
+
+array_interface_property = "__array_interface__"
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:22]: This really needs to be `__sycl_usm_array_interface__`.

File: dpctl/dparray.py (hunk 9 above)
Comment (oleksandr-pavlyk) [reply dpctl/dparray.py:122]: The check on lines 120-121 should be `hasattr(obj, array_interface_property)` where `array_interface_property` is `__sycl_usm_array_interface__` and should supersede all the subsequent checks, but keep code on lines 142-148.

File: dpctl/dparray.py (hunk 4 above)
Comment (DrTodd13) [reply dpctl/dparray.py]: Seems to be at the top now.

File: dpctl/dparray.py (hunk 9 above)
Comment (DrTodd13) [reply dpctl/dparray.py:122]: Fixed on the dpctl side.

File: dpctl/dparray.py (hunk 10 above)
Comment (DrTodd13) [reply dpctl/dparray.py:145]: Done.

File: dpctl/dparray.py (hunk 12 above)
Comment (DrTodd13) [reply dpctl/dparray.py:22]: Done.

**Comment by diptorupd** [comment]
one general comment; I want to add the new `dparray` class under `dpctl.dptensor.dparray` sub module. @oleksandr-pavlyk suggested that we can have other types of tensors defined in dpctl later and it makes sense to keep all of them in one place under a `dptensor` sub module.

File: setup.py [hunk 13]
```diff
@@ -1,25 +1,25 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
 ##
 ##===----------------------------------------------------------------------===##
 ###
-### \file
-### This file builds the dpctl and dpctl.ocldrv extension modules.
+# \file
+# This file builds the dpctl and dpctl.ocldrv extension modules.
 ##===----------------------------------------------------------------------===##
```
Comment (PokhodenkoSA) [inline setup.py:23]: I think we should use one # in all lines or do not touch it now.

Comment (PokhodenkoSA) [reply setup.py:23]: I think it is effect of black formatter. The rule is to have one # as a comment starter

File: dpctl/tests/test_dparray.py [hunk 14]
```diff
@@ -28,14 +28,6 @@ def test_multiplication_dparray(self):
...
         self.assertIsInstance(C, dparray.ndarray)
 
-    def test_dparray_through_python_func(self):
-        def func_operation_with_const(dpctl_array):
-            return dpctl_array * 2.0 + 13
-
-        C = self.X * 5
-        dp_func = func_operation_with_const(C)
-        self.assertIsInstance(dp_func, dparray.ndarray)
-
```
Comment (PokhodenkoSA) [inline dpctl/tests/test_dparray.py]: Why this test is deleted?
@vlad-perevezentsev

**Comment by PokhodenkoSA** [comment]
> add the new `dparray` class under `dpctl.dptensor.dparray` sub module

@vlad-perevezentsev Please, make it.

File: dpctl/__init__.pxd (hunk 8 above)
Comment (PokhodenkoSA) [reply dpctl/__init__.pxd]: I have removed it. Works the same.

File: setup.py (hunk 13 above)
Comment (PokhodenkoSA) [reply setup.py]: Reverted this changes.

**Review by PokhodenkoSA — COMMENTED** [review]
LGTM.

**Comment by PokhodenkoSA** [comment]
> > add the new `dparray` class under `dpctl.dptensor.dparray` sub module
> 
> @vlad-perevezentsev Please, make it in separate PR.

https://github.com/IntelPython/dpctl/pull/200

File: dpctl/dparray.py (hunk 11 above)
Comment (diptorupd) [reply dpctl/dparray.py:171]: Can we add a TODO and open a ticket to track this future work.

File: dpctl/__init__.pxd (hunk 8 above)
Comment (DrTodd13) [reply dpctl/__init__.pxd]: The testing code is referencing dpctl.dparray and it is working?  If so then perhaps we don't need this line but I'm still curious how this is working now when I had so much trouble getting it to work.

**Comment by oleksandr-pavlyk** [comment]
We should try to merge PRs #192 and #200 into a single PR before moving one, these PRs are interdependent. 

In `dpctl/dptensor/__init__.py` we should avoid dong `from .dparray import *` because it pull the entire namespace of NumPy. 
Instead we should do `from .dparray import ndarray as ndarray_USMShared`.

It would best reflect implementation is `dparray.py` is renamed to `_ndarray_subclass.py`.

Tests need to be adjusted accordingly.

Doc-strings for the new class and the new module should be added. 

I will work on `__array_ufunc__` in a separate PR.

**Comment by DrTodd13** [comment]
> We should try to merge PRs #192 and #200 into a single PR before moving one, these PRs are interdependent.
> 
> In `dpctl/dptensor/__init__.py` we should avoid dong `from .dparray import *` because it pull the entire namespace of NumPy.
> Instead we should do `from .dparray import ndarray as ndarray_USMShared`.
> 
> It would best reflect implementation is `dparray.py` is renamed to `_ndarray_subclass.py`.

I would like a name that reflects that it can be imported as a full numpy replacement.  "dparray" doesn't quite capture that but "_ndarray_subclass" I think could even more so give the wrong impression.  At some point we want to call dpnp from there so you're going to get much more than just an ndarray replacement.

File: dpctl/dparray.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+        # Just raise an exception since __array_ufunc__ makes all reasonable cases not
+        # need the code below.
+        raise ValueError(
+            "Non-USM allocated ndarray can not viewed as a USM-allocated one without a copy"
+        )
+
+    # Tell Numba to not treat this type just like a NumPy ndarray but to propagate its type.
+    # This way it will use the custom dparray allocator.
+    __numba_no_subtype_ndarray__ = True
```
Comment (diptorupd) [inline dpctl/dparray.py:161]: @DrTodd13 Do we really need this extra attribute? Can we not use the facts that the array is a subclass of NumPy.ndarray and it defines the `__sycl_usm_array_interface__` to recognize this class inside Numba?

Comment (DrTodd13) [reply dpctl/dparray.py:161]: Numba is currently made to treat Numpy.ndarray subclasses as if they were Numpy.ndarrays.  Therefore, we can't just check for whether it is a ndarray subclass to keep those types separate without breaking backwards compatibility.

Also, you might want to keep types separate even if you don't need a special allocator.  So, for generality, we shouldn't specialize on something allocator or USM-specific inside Numba.

**Comment by diptorupd** [comment]
> > We should try to merge PRs #192 and #200 into a single PR before moving one, these PRs are interdependent.
> > In `dpctl/dptensor/__init__.py` we should avoid dong `from .dparray import *` because it pull the entire namespace of NumPy.
> > Instead we should do `from .dparray import ndarray as ndarray_USMShared`.
> > It would best reflect implementation is `dparray.py` is renamed to `_ndarray_subclass.py`.
> 
> I would like a name that reflects that it can be imported as a full numpy replacement. "dparray" doesn't quite capture that but "_ndarray_subclass" I think could even more so give the wrong impression. At some point we want to call dpnp from there so you're going to get much more than just an ndarray replacement.

**Comment by PokhodenkoSA** [comment]
@diptorupd Why did you close this PR? What is the plan for it now?

**Comment by diptorupd** [comment]
@PokhodenkoSA not sure when/what I did. Must have accidentally clicked wrong button.

File: dpctl/dparray.py [hunk 16]
```diff
@@ -0,0 +1,257 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (PokhodenkoSA) [reply dpctl/dparray.py:25]: Done.

File: MANIFEST.in [hunk 17]
```diff
@@ -1,4 +1,5 @@
 include versioneer.py
+include dparray.py
```
Comment (PokhodenkoSA) [inline MANIFEST.in:2]:
```suggestion
```

**Comment by PokhodenkoSA** [comment]
> * We should try to merge PRs #192 and #200 into a single PR before moving on, these PRs are interdependent.

Not necessary. #200 could be merged later. #200 depends on #192. #200 is independent.
Now numba and numba-dppy has PR which expects dpctl.dparray exists. It should be updated too. I would like to move step by step and do not add dptensor module until we have nothing to add also.
In any case we can do in `dpctl/__init__.py` replase `from . import dparray` with `from .dptensor import dparra` and propose clinents to use `from dpctl import dparray`.
